src/extension/vscode/client.ts

```diff
@@ -0,0 +1,162 @@
+import { loadErrorCodes } from "../tab/loadErrorCodes";
+import type { MessageAdapter } from "../messageAdapters";
+import { handleExplorerRequests } from "../tab/handleExplorerRequests";
+import { setMaxListeners, WeakRef, FinalizationRegistry } from "./polyfills";
+
+type Reason =
+  | "WS_DISCONNECTED"
+  | "CLIENT_GC"
+  | "WS_ERROR"
+  | "UNREGISTER"
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  | (string & {});
+
+export function registerClient(
+  client: ApolloClient<any>,
+  url: string | URL
+): {
+  connected: Promise<void>;
+  unregister: (reason?: Reason) => void;
+  onCleanup: (cleanup: (reason: Reason) => void) => void;
+} {
+  const clientRef = new WeakRef(client);
+  const registered = _registerClient(clientRef, url);
+  const registry = new FinalizationRegistry(registered.unregister);
+  const unregisterToken = {};
+  registry.register(client, "CLIENT_GC", unregisterToken);
+  registered.onCleanup(() => registry.unregister(unregisterToken));
+  return registered;
+}
+
+function makeErrorHandler(cleanup: (reason?: Reason) => void): EventListener {
+  return (e) => {
+    console.error("Apollo DevTools Client encountered WebSocket error:", e);
+    cleanup("WS_ERROR");
+  };
+}
+
+export function _registerClient(
+  clientRef: WeakRef<ApolloClient<any>>,
+  url: string | URL
+): {
+  connected: Promise<void>;
+  unregister: (reason?: Reason) => void;
+  onCleanup: (cleanup: (reason: Reason) => void) => void;
+} {
+  const { signal, cleanup, registerCleanup } = getCleanupController();
+
+  function getClient() {
+    const client = clientRef.deref();
+    if (client) {
+      return getPrivateAccess(client);
+    } else {
+      cleanup("CLIENT_GC");
+    }
+  }
+
+  const ws = new WebSocket(url);
+  registerCleanup(ws.close.bind(ws, 1000));
+  ws.addEventListener("close", cleanup.bind(undefined, "WS_DISCONNECTED"), {
+    once: true,
+    signal,
+  });
+  ws.addEventListener("error", makeErrorHandler(cleanup), {
+    once: true,
+    signal,
+  });
+
+  const wsAdapter = createWsAdapter(ws, signal);
+  const wsRpcClient = createRpcClient(wsAdapter);
+  const wsRpcHandler = createRpcHandler(wsAdapter);
+  const wsActor = createActor(wsAdapter);
+  function getQueriesForClient() {
+    return getQueries(
+      getClient()?.queryManager.getObservableQueries("active") ?? new Map()
+    );
+  }
+  function getMutationsForClient() {
+    return getMutations(getClient()?.queryManager.mutationStore ?? {});
+  }
+  registerCleanup(
+    wsRpcHandler("getQueries", getQueriesForClient),
+    wsRpcHandler("getMutations", getMutationsForClient),
+    wsRpcHandler("getCache", () => getClient()?.cache.extract(true) ?? {}),
+    handleExplorerRequests(wsActor, getClient)
+  );
+
+  ws.addEventListener(
+    "open",
+    function open() {
+      const client = getClient();
+      if (!client) {
+        return;
+      }
+      wsActor.send({
+        type: "registerClient",
+        payload: {
+          id: "",
```

Comment:
We will need to create a unique ID per connected client. The select dropdown [uses this `id`](https://github.com/apollographql/apollo-client-devtools/blob/a413e0b8fe2c3ccab7225950439bbf6e9dccab62/src/application/App.tsx#L226) to create separate entries in the dropdown. If all clients use the same `id`, we won't be able to properly display the options in the dropdown.

Reply:
I know. Client ids are assigned in the server, which is the only instance that has the knowledge which client ids are already taken and which client id refers to which connection.

Reply:
created on the client now